extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/JpaOperations.java

```diff
@@ -393,6 +393,16 @@ public static long deleteAll(Class<?> entityClass) {
         return (long) getEntityManager().createQuery("DELETE FROM " + getEntityName(entityClass)).executeUpdate();
     }
 
+    public static boolean deleteById(Class<?> entityClass, Object id) {
```

Comment:
I think we should start with the improved version ;)

Reply:
It's not as easy as it looks, for PanacheEntity and for the repository pattern we don't know the name of the ID field so we need to retrieve it somehow ...

Reply:
I find a way to retrieve from Hiberate the name of the identifier field:
```
        EntityManager entityManager = getEntityManager();
        Session session = entityManager.unwrap(Session.class);
        SessionFactory sessionFactory = session.getSessionFactory();
        ClassMetadata metadata = sessionFactory.getClassMetadata(entityClass);
        String idField = metadata.getIdentifierPropertyName();
```
WDYT ?

Reply:
@FroMage I sucessfully create a query with the trick I described above. This couple Panache directly with Hibernate and not just JPA but I think it's OK. After all it's Hibernate with Panache not JPA with Panache :)

If the new impl is OK for you I'll squash it and we could merge it

Reply:
This looks good. Can you just check what happens with composite IDs? This should work with `@CompositeId` but not `@IdClass`, I assume.

But this should already be a limitation of `findById()` which only takes a single argument, so perhaps it's acceptable.

Reply:
https://www.baeldung.com/jpa-composite-primary-keys

Reply:
@FroMage I added tests with `@IdClass` and `@Embeddable` composite id and it works fine for `@Embeddable` (it generates a delete query with all the fields of the embeddable composite ID) but didn't works for `@IdClass`.

For `@IdClass` Hibernate didn't succeed in generating the query with the following error:
```
Caused by: java.lang.IllegalArgumentException: Could not locate ordinal parameter [1], expecting one of []
	at org.hibernate.query.internal.ParameterMetadataImpl.getOrdinalParameterDescriptor(ParameterMetadataImpl.java:154)
	at org.hibernate.query.internal.ParameterMetadataImpl.getQueryParameter(ParameterMetadataImpl.java:204)
	at org.hibernate.query.internal.QueryParameterBindingsImpl.getBinding(QueryParameterBindingsImpl.java:182)
	at org.hibernate.query.internal.AbstractProducedQuery.setParameter(AbstractProducedQuery.java:507)
	at org.hibernate.query.internal.AbstractProducedQuery.setParameter(AbstractProducedQuery.java:110)
	at io.quarkus.hibernate.orm.panache.runtime.JpaOperations.bindParameters(JpaOperations.java:95)
```

Multiple solutions here:
- It should works and I should log a bug on Hibernate side
- We go back to using `entityManager.find(id).remove()`
- We check if the object have the annotation `@IdClass` by reflection and switch back to `entityManager.find(id).remove()` only for this case (reflection lookup is fast so it should not be an issue)
- We add a varargs to deleteById so when using `@IdClass` based composite id the user can write `entity.deleteById(key.part1, key.part2)` instead of `entity.deleteById(key)` (not tested)

WDYT ?

Reply:
This stack trace is not for `findById` but for the new `deleteById` right?

Reply:
Yes, `findById works with composite id for both `@IdClass` and `@Embeddable`. Only `deleteById` with `@IdClass` didn't works

Reply:
I bet this is because we didn't find how to write a query for that.

Perhaps @Sanne knows of a way to do a delete by id in Hibernate that works without loading the entity, when using an `@IdClass`?

Reply:
"id" is a special attribute in HQL, you should be able to refer to it even if the field name is different.

Reply:
But give me some more context please; writing HQL queries for us to parse is typically not the most efficient approach.

Reply:
SERIOUSLY? I never knew that :(

The context is we're trying to implement `deleteById` without doing `findById(id).delete()`.

Reply:
@Sanne the purpose of the `deleteById` method is to ease the developement of classical REST endpoint where you want to delete an entity that are not yet loaded inside the persistence context. 

I first implement it via `findById(class, id).delete()` that basically maps to `entityManager.find(id, class); entityManager.remove(entity)`.

As this generates two queries, one SELECT and one DELETE we try to achieve a better implementation with only a DELETE query, something like : `entityManager.createQuery("delete from entity where id = ?1).setParameter(1, key).executeUpdate()`.

Keep in mind that it's inside Panache so the code is not exactly this one ...

It works, except for composite key with `@IdClass` (composite key with `@Embeddable` works fine) where it generates the exception above.

Reply:
you could do:

    Object toBeDeleted = session.byId( entityClass ).getReference( id );
    session.delete( toBeDeleted );

**However** this will not check for the object to actually exist, it will just do the delete statement.

Reply:
to clarify: the first line returns the representation of the object to be deleted in the form of a managed entity.
So if that same object was already loaded in some way, you'll get that one instance returned with the state it currently has. If there was no such object of that type and Id in the current Session, it's NOT loading it from the database but returning you a managed instance which only has the Id initialized.

- 

-

Reply:
thanks @loicmathieu , for some reason I only received your comment after posting my suggestion.
The suggestion doesn't change, just please think if the fact that you might be deleting an entity which actually no longer exists is something that you'd want an error for... I'd say probably not but I'm not 100% sure.

Reply:
@Sanne thanks a lot this seems exactly what we need ;)
We return a boolean at true if the entity was removed, I'll test it tomorrow.

Reply:
> We return a boolean at true if the entity was removed, I'll test it tomorrow.

I agree that returning this information to the user is kinda nice, but it might be very complicated to implement: Hibernate ORM won't tell you: if you have a reference to a persistent object then it will existin the DB, since we assume transactionality.

To be clear, it's not as simple as checking for the expected number of rows: it's an object so you're likely deleting what is actually a graph of entries from related tables, and this is most likely done via a combination of cascading via foreign constraints and/or batching of multiple statements into a single roundtrip.
It's very likely that it's actually cheaper to load the object first, if you need to know of its existence.

I would suggest to return `void`; if someone really needs to know, they'll always have the option to load /check explicitly. Sorry, I know it's not ideal...

Reply:
@Sanne `getReference()` performs a select in case the entity is not yet inside the persistence context and throws `EntityNotFoundException` if it didn't exist.

So we are back with a select and a delete if we use it (it didn't perform the select if the entity is already loaded inside the persistence context, but in the context of a DELETE Rest endpoint it should not be the case).

For the records, the code I tested is :
```
Object entity = getEntityManager().getReference(entityClass, id);
getEntityManager().remove(entity);
```

Reply:
@Sanne @FroMage here is what I propose:

- Implement a direct deleteByIf with HQL as previously if the identifier entity is not annotated with `@IdClass`
- Implements a deleteById via getReference + remove for entity annotated by `@IdClass`
- Document clearly the method: it should be used when the entity is not yet loaded inside the persistence context (like for stateless DELETE rest endoint), if the entity is already loaded calling `delete()` on the entity performs better.

Reply:
> @Sanne `getReference()` performs a select in case the entity is not yet inside the persistence context and throws `EntityNotFoundException` if it didn't exist.

Hi @loicmathieu , no that's not accurate. `getReference()` will not hit the database at all.

But like I warned, it will assume that you know for sure that the entity exists: it's not correct to use this method if you're in doubt.
I guess what you observe is that some subsequent operation later on is forcing the reference to be loaded - most likely the delete action itself.

TBH as I suggested previous I don't think you can implement this feature to be all three of:
 - generic enough for any model
 - correct in all such cases
 - skipping the load

It's one of those situations in which you have to pick 2.

Reply:
> it will assume that you know for sure that the entity exists: it's not correct to use this method if you're in doubt.

Yes, in fact this is my assumption also, I write my comment too quickly.

OK, so you advise to stick to `find(id) + remove()`, as it's @FroMage that ask for a JPQL implementation I would like to see if he is on the same page (I'm pretty sure yes but there is no hurry)

Reply:
If the Hibernate experts tell us to use one strategy, I won't object. Just make sure to add a comment to the impl explaining why we use what appears to be a supoptimal strategy, so that no-one wants to "improve" it later.

Reply:
yeah. We could explore some optimisations in Hibernate ORM; for example I was chatting with the team about this yesterday as I wondered why we don't do just the update statement at least for the simple cases in which we can proof there is no cascading requirements.

Turns out the team sentiment is that while that would be possible, it's actually highly unlikely for an entity to satisfy such requirements in more realistic models than our simple demos, and  possibly not trivial to do.. so that explains why it doesn't do it but there certainly is the option to eventually improve on that.

Reply:
BTW another angle to this is that even if we *did* (hypotethically speaking) implement such an optimisation in ORM, you'd have no way to return true/false about knowing if the specific entity was actually existing before... so probably really not worth it.